### PR TITLE
Add lazy checkpoint loading for FSDP full-state checkpoints

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Addresses (2) in https://github.com/Lightning-AI/lightning/issues/18008#issuecomment-1627177014
Closes #18138

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @borda @justusschock @awaelchli @carmocca

### CODE REVIEW

File: src/lightning/fabric/strategies/fsdp.py [hunk 1]
```diff
@@ -821,9 +827,8 @@ def _load_raw_module_state(path_or_ckpt: Union[Path, Dict[str, Any]], module: Mo
 
     from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
 
-    # This is inefficient, as multiple copies of the checkpoint are held in CPU memory at once.
-    # There is currently no other way because `summon_full_params` does not support write-back from rank 0 only.
-    state_dict = torch.load(path_or_ckpt, map_location="cpu") if not isinstance(path_or_ckpt, dict) else path_or_ckpt
+    # Use `lazy_load` instead of `torch.load` here to avoid storing a copy of the full checkpoint per rank
+    state_dict = _lazy_load(path_or_ckpt) if isinstance(path_or_ckpt, Path) else path_or_ckpt
```
Comment (carmocca) [inline src/lightning/fabric/strategies/fsdp.py:831]: Where do we pass `path_or_ckpt` as a path? I hadn't noticed it before but I find it strange to allow both options.

File: src/lightning/fabric/utilities/load.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        if module == "torch._utils" and name == "_rebuild_parameter":
+            return partial(_NotYetLoadedTensor.rebuild_parameter, archiveinfo=self)
+        return super().find_class(module, name)
+
+    def persistent_load(self, pid: tuple) -> "TypedStorage":
+        from torch.storage import TypedStorage
+
+        name, cls, fn, device, size = pid
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore")  # TODO: needed?
```
Comment (carmocca) [inline src/lightning/fabric/utilities/load.py:184]: The TypedStorage APIs have heavy deprecations in torch. This is okay with me since it's wrapped by `catch_warnings`

**Review by carmocca — APPROVED** [review]
Awesome! Have you thought about other places in the codebase where we could lazy load?

We can now import our utility from lit-gpt

**Comment by awaelchli** [comment]
> Have you thought about other places in the codebase where we could lazy load?

It looks like it is difficult to integrate it into the rest of the code base. Several things are standing in the way that require us to brainstorm on design decisions. Here are some raw notes from my notebook:

- Lazy tensors can't be loaded by an optimizer, because the optimizer creates a deepcopy, causing pickling issues.
- Lazy load bypasses all the torch.load logic, which means for example it doesn't support map_location.
- Loading from filsystems via fsspec is not supported
- It is hard to integrate lazy load with the existing IO plugin structure, since it needs to be injected at the lowest level where the file gets read.
- Tensors remain lazy after loading (e.g. on_load_checkpoint hook), the user can't access the weights unless he materializes them manually.

Happy to open an issue about this for further discussion!

> We can now import our utility from lit-gpt

There might be slight differences between the version here and in lit-gpt. I need to double check, there was something about quantization.

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (awaelchli) [reply src/lightning/fabric/strategies/fsdp.py:831]: Happy to refactor this in a follow-up. The logic has quickly grown over the last few weeks and we need to take care of it, I agree.

Two places where we call it:
https://github.com/Lightning-AI/lightning/blob/6511ac28759718a524dd00e627c186fb6baea763/src/lightning/fabric/strategies/fsdp.py#L510-L512

https://github.com/Lightning-AI/lightning/blob/6511ac28759718a524dd00e627c186fb6baea763/src/lightning/fabric/strategies/fsdp.py#L575-L577

File: src/lightning/fabric/utilities/load.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+            "size",
+            "shape",
+            "volatile",
+        }:
+            return getattr(self.metatensor, name)
+
+        # TODO: needed for us?
+        # materializing with contiguous is needed for quantization
+        if name in {"contiguous"}:
+            return getattr(self._load_tensor(), name)
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/load.py:157]: I should probably keep this here if we want to import and use the util in lit-gpt?

File: src/lightning/fabric/utilities/load.py (hunk 2 above)
Comment (awaelchli) [reply src/lightning/fabric/utilities/load.py]: After your comment, I added a comment to the code that this is the reason why we suppress warnings

File: src/lightning/fabric/strategies/fsdp.py (hunk 1 above)
Comment (carmocca) [reply src/lightning/fabric/strategies/fsdp.py:831]: Sounds good to me

File: src/lightning/fabric/utilities/load.py (hunk 3 above)
Comment (carmocca) [reply src/lightning/fabric/utilities/load.py]: It might also be a generally applicable limitation. On the other hand, we could do this manually in the quantization code directly.

What do you suggest @t-vi?

**Review by Borda — APPROVED** [review]
